### PR TITLE
Ability to blacklist ip ranges for federation traffic

### PR DESCRIPTION
**Pull request by anoadragon453**
Closes https://github.com/matrix-org/synapse/issues/3953

This PR adds a new config option that functions similar to `url_preview_ip_range_blacklist`, but provides a blacklist for federation traffic. It does this by making use of our `IPBlacklistResolver` and `BlacklistingAgentWrapper` classes.

Sytest PR: https://github.com/matrix-org/sytest/pull/607

### CODE REVIEW

File: synapse/config/server.py [hunk 1]
```diff
@@ -351,6 +375,20 @@ def default_config(self, server_name, data_dir_path, **kwargs):
         #  - nyc.example.com
         #  - syd.example.com
 
+        # Prevent federation requests from being sent to the following
+        # blacklist IP address CIDR ranges.
```
Comment (richvdh) [inline synapse/config/server.py]: please could you explain what the default is (even if it's that nothing is blacklisted)?

File: synapse/config/server.py [hunk 2]
```diff
@@ -351,6 +375,20 @@ def default_config(self, server_name, data_dir_path, **kwargs):
...
+        #
+        #federation_ip_range_blacklist:
```
Comment (richvdh) [inline synapse/config/server.py]: do you think we should uncomment this so that we blacklist the RFC1918 addresses by default for new deployments?

File: synapse/http/client.py [hunk 3]
```diff
@@ -109,7 +111,12 @@ def addressResolved(address):
                     logger.info(
                         "Dropped %s from DNS resolution to %s" % (ip_address, hostname)
                     )
-                    raise SynapseError(403, "IP address blocked by IP blacklist entry")
+                    # Only raise a 403 if this request originated from a
```
Comment (richvdh) [inline synapse/http/client.py]: why do we need to do this?

File: synapse/config/server.py [hunk 4]
```diff
@@ -122,6 +122,30 @@ def read_config(self, config):
             for domain in federation_domain_whitelist:
                 self.federation_domain_whitelist[domain] = True
 
+        self.federation_ip_range_blacklist = config.get(
+            "federation_ip_range_blacklist", None,
```
Comment (richvdh) [inline synapse/config/server.py]: If we make the default `[]` rather than `None` we can save ourselves a bunch of special-casing?

File: tests/http/test_fedclient.py [hunk 5]
```diff
@@ -78,7 +78,7 @@ def do_request():
         # Nothing happened yet
         self.assertNoResult(test_d)
 
-        # Make sure treq is trying to connect
+        # Make sure the req is trying to connect
```
Comment (richvdh) [inline tests/http/test_fedclient.py]: `treq` is correct

File: tests/http/test_fedclient.py [hunk 6]
```diff
@@ -211,6 +211,110 @@ def test_client_connect_no_response(self):
         self.assertIsInstance(f.value, RequestSendFailed)
         self.assertIsInstance(f.value.inner_exception, ResponseNeverReceived)
 
+    def test_client_ip_range_blacklist(self):
+        """Ensure that Synapse does not try to connect to blacklisted IPs"""
+        # Set up the ip_range blacklist
+        from netaddr import IPSet
```
Comment (richvdh) [inline tests/http/test_fedclient.py]: import this at the top?

File: synapse/config/server.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/config/server.py]: Seeing as it would be a security benefit for the vast majority of use cases, I would argue yes.

File: synapse/http/client.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/http/client.py]: I'm struggling to remember now. I remember I worked with @erikjohnston on this.

Something will passing tests perhaps?

File: synapse/config/server.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/config/server.py]: I suppose so since this is a blacklist and an empty one is the same as not having one at all.

File: synapse/http/client.py (hunk 3 above)
Comment (anoadragon453) [reply synapse/http/client.py]: T'was actually due to special-casing for unit tests, but we've now removed this, as the existing implementation didn't make sense anyways.

File: tests/http/test_fedclient.py [hunk 7]
```diff
@@ -211,6 +211,110 @@ def test_client_connect_no_response(self):
...
+                    fetch_res = yield fetch_d
+                    defer.returnValue(fetch_res)
+                finally:
+                    check_logcontext(context)
+
+        # Make the request
+        d = do_request()
+        self.pump()
+
+        # Nothing has happened yet
```
Comment (richvdh) [inline tests/http/test_fedclient.py]: please remember to test that these deferreds eventually get resolved one way or the other, otherwise we'll end up with requests blocking forever.

Comment (anoadragon453) [reply tests/http/test_fedclient.py]: Still not 100% sure on how to do this. `yield`'ing results in the function timing out (though maybe that's what we want to test?).

File: synapse/config/server.py (hunk 2 above)
Comment (richvdh) [reply synapse/config/server.py]: so would I. Please can you make it so?

File: synapse/http/matrixfederationclient.py [hunk 8]
```diff
@@ -172,19 +175,43 @@ def __init__(self, hs, tls_client_options_factory):
...
+
+        self.reactor = Reactor()
 
         self.agent = MatrixFederationAgent(
-            hs.get_reactor(),
+            self.reactor,
             tls_client_options_factory,
         )
+
+        # Prevent direct connections to blacklisted IP addresses
```
Comment (richvdh) [inline synapse/http/matrixfederationclient.py]: I don't find "direct connections" very clear here. How about:

"Use a BlacklistingAgentWrapper to prevent circumventing the IP blacklist via IP literals in server names"

File: tests/http/test_fedclient.py (hunk 7 above)
Comment (richvdh) [reply tests/http/test_fedclient.py]: > `yield`'ing results in the function timing out

yes. The thing to note here is that the test function itself is running *synchronously*: it starts at the beginning, and doesn't return until it is finished. Note that is different to a `Deferred`-returning function, which only does part of its work synchronously, and leaves the rest to be done later (at which point, the `Deferred` will resolve).

Now, the way that we test the asynchronous code from a synchronous function is something like:

1. run a function that returns a deferred
2. satisfy the condition that the function is waiting for (eg: pretend that some time has passed; connect a TCP socket; spoof some results from the database).
  * Once we satisfy the awaited condition, the second half of the function should complete, and hence resolve its deferred.
3. Check that the deferred has now resolved with the expected result.

Often, all that is needed is to "pretend some time has passed". We do this by using a fake Reactor where we can manipulate its idea of the current time. "Pretending time has passed" is then called "advancing the reactor", and is what `self.pump` does.  

[It may be helpful to note that this means that there are actually *two* reactors in play, since the test runner (`trial`) is a Twisted application which runs on a regular reactor: so we end up with the main reactor which is running `trial` and the test reactor which is running the code under test. But in practice, as long as your test functions are synchronous, you can forget about the `trial` reactor.]

So, long story short: how do we test that the deferreds eventually get resolved? well, somewhere in the test we ought to be able to call `self.successResultOf(d)` or `self.failureResultOf(d, cls)` [iirc]. If that assertion fails, it means that the deferred isn't completing, and that is a bad thing!

File: synapse/config/server.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/config/server.py]: Done so, but requires having sytest override it so that tests don't fail.

File: synapse/config/server.py [hunk 9]
```diff
@@ -124,6 +126,21 @@ def read_config(self, config):
             for domain in federation_domain_whitelist:
                 self.federation_domain_whitelist[domain] = True
 
+        self.federation_ip_range_blacklist = config.get(
```
Comment (richvdh) [inline synapse/config/server.py]: we ought to follow the example of #5134 and include 0.0.0.0 and `::`, whether they were explicitly listed or not.

File: synapse/http/client.py [hunk 10]
```diff
@@ -90,45 +90,50 @@ def __init__(self, reactor, ip_whitelist, ip_blacklist):
     def resolveHostName(self, recv, hostname, portNumber=0):
 
         r = recv()
-        d = defer.Deferred()
```
Comment (richvdh) [inline synapse/http/client.py]: can you merge #5155 so that we can get this stuff out of the diff?

File: tests/http/test_fedclient.py [hunk 11]
```diff
@@ -211,6 +213,75 @@ def test_client_connect_no_response(self):
...
+
+        # Try making a GET request to a blacklisted IPv4 address
+        # ------------------------------------------------------
+        # Make the request
+        d = cl.get_json("internal:8008", "foo/bar", timeout=10000)
+
+        # Nothing happened yet
+        self.assertNoResult(d)
+
+        self.pump(120)
```
Comment (richvdh) [inline tests/http/test_fedclient.py]: hrm, having to wait 120 seconds for this doesn't sound right. Why does it take so long?

Comment (anoadragon453) [reply tests/http/test_fedclient.py]: Oh, it doesn't. I've just selected that value arbitrarily.

File: synapse/config/server.py [hunk 12]
```diff
@@ -148,6 +148,9 @@ def read_config(self, config):
             self.federation_ip_range_blacklist = IPSet(
                 self.federation_ip_range_blacklist
             )
+
+            # Always blacklist 0.0.0.0, ::
+            self.federation_ip_range_blacklist.update(["0.0.0.0", "::"])
```
Comment (richvdh) [inline synapse/config/server.py]: this could really have done with being outside the try/catch, but nm
